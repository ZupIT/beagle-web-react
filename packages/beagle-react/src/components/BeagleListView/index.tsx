/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import React, { FC, useEffect, useRef, Children } from 'react'
import { BeagleUIElement } from '@zup-it/beagle-web'
import { Tree, logger } from '@zup-it/beagle-web'
import withTheme from '../utils/withTheme'
import useScroll from './scroll'
import { StyledListView } from './styled'
import { BeagleListViewInterface } from './types'

const BeagleListView: FC<BeagleListViewInterface> = ({
  direction = 'VERTICAL',
  className,
  style,
  template,
  onInit,
  onScrollEnd,
  scrollEndThreshold = 100,
  dataSource,
  iteratorName = 'item',
  viewContentManager,
  children,
  useParentScroll = false,
  _key,
  __suffix__,
}) => {
  const elementRef = useRef() as React.MutableRefObject<HTMLDivElement>
  const hasRendered = !Array.isArray(dataSource) || dataSource.length === Children.count(children)
  useScroll(
    { elementRef, direction, onScrollEnd, scrollEndThreshold, useParentScroll, hasRendered },
    [Children.count(children)],
  )

  useEffect(() => {
    if (onInit) onInit()
  }, [])

  useEffect(() => {
    if (!Array.isArray(dataSource)) return

    if (!viewContentManager) {
      return logger.error('The beagle:listView component should only be used inside a view rendered by Beagle.')
    }

    const element = viewContentManager.getElement() as BeagleUIElement
    if (!element) return
    const listViewTag = viewContentManager.getElement()._beagleComponent_.toLowerCase()
    const listViewId = viewContentManager.getElement().id

    element.children = dataSource.map((item, index) => {
      const templateTree = Tree.clone(template)
      const iterationKey = _key && item[_key] !== undefined ? item[_key] : index
      const suffix = __suffix__ || ''
      templateTree._implicitContexts_ = [{ id: iteratorName, value: item }]
      Tree.forEach(templateTree, (component, componentIndex) => {
        const baseId = component.id ? `${component.id}${suffix}` : `${listViewId}:${componentIndex}`
        component.id = `${baseId}:${iterationKey}`
        if (component._beagleComponent_.toLowerCase() === listViewTag) {
          component.__suffix__ = `${suffix}:${iterationKey}`
        }
      })
      
      return templateTree
    })

    viewContentManager.getView().getRenderer().doFullRender(element, element.id)
  }, [JSON.stringify(dataSource)])

  return (
    <StyledListView
      ref={elementRef}
      className={className}
      direction={direction}
      useParentScroll={useParentScroll}
      style={style}
    >
      {children}
    </StyledListView>
  )
}

export default withTheme(BeagleListView)
