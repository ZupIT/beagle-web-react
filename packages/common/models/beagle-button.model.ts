import { ClickEvent } from '@zup-it/beagle-web'
import { BeagleComponent } from '../types'
import { BeagleDefaultComponent } from './types'


export interface BeagleButtonInterface extends BeagleDefaultComponent, BeagleComponent {
  text: string,
  onPress?: () => void,
  clickAnalyticsEvent?: ClickEvent,
  disabled?: boolean, 
}
