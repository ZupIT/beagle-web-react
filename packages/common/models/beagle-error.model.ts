import { ErrorComponentParams } from '@zup-it/beagle-web'
import { BeagleDefaultComponent } from './types'

export interface BeagleErrorInterface extends BeagleDefaultComponent, ErrorComponentParams {
}
