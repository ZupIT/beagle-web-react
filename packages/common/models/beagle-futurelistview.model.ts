/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { BeagleUIElement } from '@zup-it/beagle-web'
import { BeagleComponent } from '../types'
import { Direction, BeagleDefaultComponent } from './types'

export type NodeType = HTMLElement | null

export interface BeagleFutureListViewInterface extends BeagleDefaultComponent, BeagleComponent {
  direction: Direction,
  dataSource: any[],
  iteratorName?: string,
  onInit?: () => void,
  onScrollEnd?: () => void,
  scrollEndThreshold?: number,
  template: BeagleUIElement,
  useParentScroll?: boolean,
  /* the property "key" is not allowed in React. Since this is not a rule for Beagle, every time
  Beagle receives "key", it transforms it into "_key" */
  _key?: string,
  __suffix__?: string,
}

export interface ScrollInterface {
  direction: Direction,
  onScrollEnd?: () => void,
  scrollEndThreshold: number,
  useParentScroll: boolean,
  elementRef: React.MutableRefObject<HTMLDivElement>,
  hasRendered: boolean,
}
