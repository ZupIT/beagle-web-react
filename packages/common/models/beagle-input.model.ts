import { BeagleDefaultComponent } from './types'

export type InputHandler = (event: { value: string }) => void

export type InputType = 'DATE' | 'EMAIL' | 'PASSWORD' | 'NUMBER' | 'TEXT'

export interface BeagleTextInputInterface extends BeagleDefaultComponent {
  value: string,
  onChange?: InputHandler,
  onFocus?: InputHandler,
  onBlur?: InputHandler,
  /**
  * @deprecated since version 1.7.0. Use `enabled` instead.
  *  Use field `enabled` to control whether the button is enabled or not in the layout.
  */
  disabled?: boolean,
  enabled?: boolean,
  readOnly?: boolean,
  placeholder?: string,
  type?: InputType,
  error?: string,
  showError?: boolean,
}

export interface BeagleTextAreaInterface extends BeagleTextInputInterface {
  name?: string,
  label?: string,
}

