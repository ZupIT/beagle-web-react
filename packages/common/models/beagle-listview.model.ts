import { BeagleDefaultComponent, Direction } from './types'

export interface BeagleListViewInterface extends BeagleDefaultComponent {
  direction: Direction,
}
