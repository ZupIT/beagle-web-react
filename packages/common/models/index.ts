/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/
export * from './beagle-image.model'
export * from './beagle-text.model'
export * from './beagle-button.model'
export * from './beagle-listview.model'
export * from './beagle-modal.model'
export * from './beagle-input.model'
export * from './beagle-container.model'
export * from './beagle-touchable.model'
export * from './beagle-tabbar.model'
export * from './beagle-pageview.model'
export * from './beagle-pageindicator.model'
export * from './beagle-lazy.model'
export * from './beagle-simpleform.model'
export * from './beagle-webview.model'
export * from './beagle-futurelistview.model'
export * from './beagle-screen.model'
export * from './beagle-error.model'
export * from './types'
