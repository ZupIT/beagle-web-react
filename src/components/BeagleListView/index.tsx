/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import React, { FC } from 'react'
import { Direction } from '../types'
import { StyledListView } from './styled'
  
export interface BeagleListViewInterface {
  direction: Direction,
  className?: string,
  style?: React.CSSProperties,
}

const BeagleListView: FC<BeagleListViewInterface> = props => {
  const { children, direction, className, style } = props

  return (
    <StyledListView className={className} direction={direction} style={style}>
      {children}
    </StyledListView>
  )
}

export default BeagleListView
