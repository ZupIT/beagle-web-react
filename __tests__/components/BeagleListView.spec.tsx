/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Link.react.test.js
import React from 'react'
import 'jest-styled-components'
import Adapter from 'enzyme-adapter-react-16'
import { configure, shallow } from 'enzyme'
import BeagleListView from 'components/BeagleDynamicLists/BeagleListView'
import { Direction } from 'models'

let wrapper: any
const directionMock: Direction = 'HORIZONTAL'
const mockStyle: React.CSSProperties = {
  height: '100',
  width: '50',
}

configure({ adapter: new Adapter() })

describe('Render BeagleListView with multiple templates', () => {
  const dataSource = [
    { key: 'first', text: 'First Child' },
    { key: 'second', text: 'Second Child' },
    { key: 'third', text: 'First Child' },
    { key: 'unknown', text: 'Unknown Child' },
  ]

  const templates = [
    {
      case: "@{eq(item.key, 'first')}",
      view: {
        _beagleComponent_: 'beagle:container',
        id: `first-child`,
        children: [
          {
            _beagleComponent_: 'beagle:textInput',
            id: `first-text`,
            value: 'first',
          }
        ],
      },
    },
    {
      case: "@{eq(item.key, 'second')}",
      view: {
        _beagleComponent_: 'beagle:textInput',
        id: `second-child`,
        value: 'second',
      },
    },
    {
      case: "@{eq(item.key, 'third')}",
      view: {
        _beagleComponent_: 'beagle:container',
        id: `third-child`,
        children: [
          {
            _beagleComponent_: 'beagle:textInput',
            id: `third-label`,
            value: `third label`,
          },
          {
            _beagleComponent_: 'beagle:input',
            id: `third-input`,
            value: `third input`,
          }
        ],
      },
    },
    {
      view: {
        _beagleComponent_: 'beagle:textInput',
        id: `default-child`,
        value: 'default',
      },
    },
  ]

  beforeAll(() => {
    wrapper = shallow(
      <BeagleListView
        direction={directionMock}
        style={mockStyle}
        dataSource={dataSource}
        templates={templates}
        className="Test Class"
      />
    )
  })
  
  test('Beagle snapshot list view', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
