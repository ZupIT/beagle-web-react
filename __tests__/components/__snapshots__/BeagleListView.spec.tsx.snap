// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render BeagleListView with multiple templates Beagle snapshot list view 1`] = `
<ComponentWithTheme
  className="Test Class"
  dataSource={
    Array [
      Object {
        "key": "first",
        "text": "First Child",
      },
      Object {
        "key": "second",
        "text": "Second Child",
      },
      Object {
        "key": "third",
        "text": "First Child",
      },
      Object {
        "key": "unknown",
        "text": "Unknown Child",
      },
    ]
  }
  direction="HORIZONTAL"
  listType="LIST"
  style={
    Object {
      "height": "100",
      "width": "50",
    }
  }
  templates={
    Array [
      Object {
        "case": "@{eq(item.key, 'first')}",
        "view": Object {
          "_beagleComponent_": "beagle:container",
          "children": Array [
            Object {
              "_beagleComponent_": "beagle:textInput",
              "id": "first-text",
              "value": "first",
            },
          ],
          "id": "first-child",
        },
      },
      Object {
        "case": "@{eq(item.key, 'second')}",
        "view": Object {
          "_beagleComponent_": "beagle:textInput",
          "id": "second-child",
          "value": "second",
        },
      },
      Object {
        "case": "@{eq(item.key, 'third')}",
        "view": Object {
          "_beagleComponent_": "beagle:container",
          "children": Array [
            Object {
              "_beagleComponent_": "beagle:textInput",
              "id": "third-label",
              "value": "third label",
            },
            Object {
              "_beagleComponent_": "beagle:input",
              "id": "third-input",
              "value": "third input",
            },
          ],
          "id": "third-child",
        },
      },
      Object {
        "view": Object {
          "_beagleComponent_": "beagle:textInput",
          "id": "default-child",
          "value": "default",
        },
      },
    ]
  }
/>
`;
